File: src/programmers/internal/json_schema_escaped.ts
import { OpenApi, OpenApiTypeChecker } from "@samchon/openapi";

import { Metadata } from "../../schemas/metadata/Metadata";
import { MetadataEscaped } from "../../schemas/metadata/MetadataEscaped";

import { json_schema_station } from "./json_schema_station";

/**
 * @internal
 */
export const json_schema_escaped = (props: {
  components: OpenApi.IComponents;
  escaped: MetadataEscaped;
}): OpenApi.IJsonSchema[] => {
  const output: OpenApi.IJsonSchema | null = json_schema_station({
    blockNever: false,
    components: props.components,
    metadata: props.escaped.returns,
    attribute: {},
  });
  if (output === null) return [];

  if (
    is_date({
      visited: new Set(),
      metadata: props.escaped.original,
    })
  ) {
    const string: OpenApi.IJsonSchema.IString | undefined =
      OpenApiTypeChecker.isString(output)
        ? output
        : OpenApiTypeChecker.isOneOf(output)
          ? (output.oneOf.find(
              OpenApiTypeChecker.isString,
            ) as OpenApi.IJsonSchema.IString)
          : undefined;
    if (
      string !== undefined &&
      string.format !== "date" &&
      string.format !== "date-time"
    )
      string.format = "date-time";
  }
  return OpenApiTypeChecker.isOneOf(output)
    ? (output.oneOf as OpenApi.IJsonSchema[])
    : [output];
};

/**
 * @internal
 */
const is_date = (props: {
  visited: Set<Metadata>;
  metadata: Metadata;
}): boolean => {
  if (props.visited.has(props.metadata)) return false;
  props.visited.add(props.metadata);

  return (
    props.metadata.natives.some((native) => native.name === "Date") ||
    props.metadata.arrays.some((array) =>
      is_date({
        visited: props.visited,
        metadata: array.type.value,
      }),
    ) ||
    props.metadata.tuples.some((tuple) =>
      tuple.type.elements.some((e) =>
        is_date({
          visited: props.visited,
          metadata: e,
        }),
      ),
    ) ||
    props.metadata.aliases.some((alias) =>
      is_date({
        visited: props.visited,
        metadata: alias.type.value,
      }),
    )
  );
};

File: packages/typia/native/core/programmers/iterate/json_schema_escaped.go
package iterate

import nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"

func json_schema_escaped(props struct {
  components *OpenApi_IComponents
  escaped    *nativemetadata.MetadataEscaped
}) []JsonSchema {
  output := json_schema_station(json_schema_station_props{
    blockNever: false,
    components: props.components,
    metadata:   props.escaped.Returns,
    attribute:  JsonSchema{},
  })
  if output == nil {
    return []JsonSchema{}
  }
  if json_schema_escaped_is_date(map[*nativemetadata.MetadataSchema]bool{}, props.escaped.Original) {
    if json_schema_is_string(output) {
      if output["format"] != "date" && output["format"] != "date-time" {
        output["format"] = "date-time"
      }
    } else if oneOf, ok := output["oneOf"].([]JsonSchema); ok {
      for _, schema := range oneOf {
        if json_schema_is_string(schema) {
          if schema["format"] != "date" && schema["format"] != "date-time" {
            schema["format"] = "date-time"
          }
          break
        }
      }
    }
  }
  if oneOf, ok := output["oneOf"].([]JsonSchema); ok {
    return oneOf
  }
  return []JsonSchema{output}
}

func json_schema_escaped_is_date(visited map[*nativemetadata.MetadataSchema]bool, metadata *nativemetadata.MetadataSchema) bool {
  if visited[metadata] {
    return false
  }
  visited[metadata] = true
  for _, native := range metadata.Natives {
    if native.Name == "Date" {
      return true
    }
  }
  for _, array := range metadata.Arrays {
    if json_schema_escaped_is_date(visited, array.Type.Value) {
      return true
    }
  }
  for _, tuple := range metadata.Tuples {
    for _, elem := range tuple.Type.Elements {
      if json_schema_escaped_is_date(visited, elem) {
        return true
      }
    }
  }
  for _, alias := range metadata.Aliases {
    if json_schema_escaped_is_date(visited, alias.Type.Value) {
      return true
    }
  }
  return false
}
